src/timeutils.cc

```diff
@@ -0,0 +1,184 @@
+            res->tv_nsec = static_cast<long>(nanosecdbl) % 1000000000;
+            res->tv_sec = static_cast<long>(nanosecdbl) / 1000000000;
+            return 0;
+        }
+    }
+    return -1;
+}
+
+int clock_getres(clockid_t clk_id, struct timespec *res) {
+    if (clk_id == CLOCK_MONOTONIC)
+        return clock_getres_monotonic(res);
+    return -1;
+}
+
+int clock_gettime_monotonic(struct timespec *ts) {
+    static bool bInitDone = false;
+    if (ts == nullptr)
+        return -1;
+
+    LARGE_INTEGER           t;
+//	FILETIME            f;
+    double                  nanoseconds;
+    static LARGE_INTEGER    starttime;
+    static double           TicksPerNanosecond;
+
+    if (!bInitDone) {
+        QueryPerformanceCounter(&starttime);
+
+        LARGE_INTEGER performanceFrequency;
+        QueryPerformanceFrequency(&performanceFrequency);
+        bInitDone = true;
+        TicksPerNanosecond = (double)performanceFrequency.QuadPart / 1000000000.;
+
+    }
+
+    QueryPerformanceCounter(&t);
+    t.QuadPart -= starttime.QuadPart;
+
+    nanoseconds = (double)t.QuadPart / TicksPerNanosecond;
+    ts->tv_nsec = static_cast<long>(nanoseconds) % 1000000000;
+    ts->tv_sec = static_cast<long>(nanoseconds) / 1000000000;
+
+    return 0;
+}
+
+/* FILETIME for January 1st, 1970 */
+unsigned __int64 jan_1_1970 = static_cast<unsigned __int64>(116444736000000000ULL);
+
+int clock_gettime_realtime(struct timespec *ts) {
+    int retval = -1;
+    if (ts != nullptr) {
+        FILETIME    filetime;
+        SYSTEMTIME    systemtime;
+        ULARGE_INTEGER ulI;
+
+        GetSystemTimeAsFileTime(&filetime);
+        FileTimeToSystemTime(&filetime, &systemtime);
+
+        ulI.LowPart = filetime.dwLowDateTime;
+        ulI.HighPart = filetime.dwHighDateTime;
+
+        ts->tv_sec = static_cast<long>((ulI.QuadPart - jan_1_1970) / 10000000L);
+        ts->tv_nsec = static_cast<long>(systemtime.wMilliseconds * 1000000);
+        retval = 0;
+    }
+    return retval;
+
+
+}
+
+int clock_gettime(clockid_t clk_id, struct timespec *ts) {
+    if (clk_id == CLOCK_MONOTONIC)
+        return clock_gettime_monotonic(ts);
+    else if (clk_id == CLOCK_REALTIME)
+        return clock_gettime_realtime(ts);
+    return -1;
+
+}
+
+LARGE_INTEGER getFILETIMEoffset() {
+    SYSTEMTIME s;
+    FILETIME f;
+    LARGE_INTEGER t;
+
+    s.wYear = 1970;
+    s.wMonth = 1;
+    s.wDay = 1;
+    s.wHour = 0;
+    s.wMinute = 0;
+    s.wSecond = 0;
+    s.wMilliseconds = 0;
+    SystemTimeToFileTime(&s, &f);
+    t.QuadPart = f.dwHighDateTime;
+    t.QuadPart <<= 32;
+    t.QuadPart |= f.dwLowDateTime;
+    return (t);
+}
+
+
+char *ctime_r(const time_t *timep, char buf[]) {
+    if (timep == nullptr || buf == nullptr)
+        return nullptr;
+    char buffer[100];
+    errno_t e = ctime_s(buffer, 100, timep);
+    if (e == 0)
+    {
+#pragma warning(disable:4996)
+        //assuming buf has enough space
+        strcpy(buf, buffer);
```

Comment:
Please, use `http://www.cplusplus.com/reference/cstring/strncpy/` here. This will enable you to remove that assumption and warning disable.